Ensure that if the array length is less than the buffer lenght we get…

AlexGhiondea opened the pull request:
… the right sized buffer

Fixes #9022

AlexGhiondea committed:
Ensure that if the array length is less than the buffer lenght we get the right sized buffer

pakrym commented:
Test? :)

AlexGhiondea committed:
Remove dead code and add basic tests.

pakrym commented on sdk/core/Azure.Core/tests/AzureBaseBuffersExtensionsTests.cs:
```diff
@@ -0,0 +1,75 @@
+    public class AzureBaseBuffersExtensionsTests
+    {
+        [Test]
+        public async Task WriteBufferToStream()
```
I don't think you are exercising the `!MemoryMarshal.TryGetArray` here because in all the cases memory is backed by the array.

You might need to implement a `MemoryManager` to hit it.

AlexGhiondea replied:
Yeah, I saw that. I did verify this under the debugger though. 

As for writing a `MemoryManager` -- even in that case I am not sure if that would be enough as we are using the `MemoryMarshal` underneath and I haven't figured out a good way of replacing the manager insider that class.

pakrym replied:
https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.IO/tests/Stream/Stream.ReadWriteSpan.cs#L116

KrzysztofCwalina commented on sdk/core/Azure.Core/src/Internal/AzureBaseBuffersExtensions.cs:
```diff
@@ -27,12 +27,8 @@ public static async Task WriteAsync(this Stream stream, ReadOnlyMemory<byte> buf
+                    array = ArrayPool<byte>.Shared.Rent(buffer.Length);
+
                     if (!buffer.TryCopyTo(array))
                         throw new Exception("could not rent large enough buffer.");
```
I know this is not part of the PR, but can ArrayPool.Rent return buffer shorter than requested? If no, then this exception is not good. If yes, then instead throwing, we should either allocate a buffer on the heap or borrow a shorter buffer and copy the data in chunks.

pakrym replied:
AFAIK, it can't be.

pakrym replied:
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs#L144

KrzysztofCwalina replied:
hopefully this will never change.

AlexGhiondea replied:
In this case, the exception will probably not going to be thrown. In the sequence case there was another bug that would have led to the exception thrown (because of a wrong check).

AlexGhiondea committed:
Address an additional bug in the extension method.

When the size of the current segment was larger than the size of the local array we should have returned the array and get a larger array.
However, the check was looking at the entire size of the buffer and comparing that to the lenght of the current segment -- this would always be false.

Without the change, we would have thrown an exception when writing a sequence in which the current segment was larger than the previous one.

Added tests for both array-backed buffers as well as for NativeMemory backed ones. This required that the Azure.Core.Tests project builds with /unsafe option.

AlexGhiondea commented on sdk/core/Azure.Core/src/Internal/AzureBaseBuffersExtensions.cs:
```diff
@@ -63,7 +59,7 @@ public static async Task WriteAsync(this Stream stream, ReadOnlySequence<byte> b
                     else
                     {
-                        if (array == null || buffer.Length < segment.Length)
+                        if (array == null || array.Length < segment.Length)
```
We need to check the `array.Length` here since the `array` holds the existing segment in it. 'buffer.Length' is the length of the entire sequence which will never be smaller than the length of an individual segment (it can be equal).

AlexGhiondea committed:
Add one more test to validate the case when both sides of the if statement are executed in the same loop.

AlexGhiondea committed:
Remove extra usings/
